Decouple `kani` version from `kani-github-action` version

jaisnan opened the pull request:
### Description of changes: 

This PR updates the action to add the parameter for `kani-version` which defaults to the value latest.
The default value latest uses the latest version of `Kani` that is uploaded on `crates.io` unless a specific version is provided by the user, in which case that specific version of `Kani` will be used in the CI.

Documentation changes [rendered here.](https://github.com/jaisnan/kani-github-action/tree/use-kani-latest-by-default)

The changes were tested in a local setting and a test project here:

1. Test for [using "latest" by default](https://github.com/jaisnan/test-vscode-extension/actions/runs/6101293421/job/16559528415). Related kani-github-action job [here](https://github.com/jaisnan/test-vscode-extension/blob/f0afa13b4542157d27c2aae9cc5a93a5e0dca4a2/.github/workflows/ci.yaml).
2. Test for [using pinned kani version](https://github.com/jaisnan/test-vscode-extension/actions/runs/6102054828/job/16559681373?pr=2). Related kani-github-action job [here](https://github.com/jaisnan/test-vscode-extension/blob/2e04619cd75bb2692cbdccba791144a000aafd65/.github/workflows/ci.yaml). 

Resolves: https://github.com/model-checking/kani-github-action/issues/46

### Call-outs:

This would update the action to pull in the latest version by default. Users who don't provide a value for the kani-version parameter would have to set it to pin the `kani` version that they want to use.

### Testing:

* How is this change tested? Tested locally, Added a CI check for the `kani-version` parameter, and ran the feature branch of the action on a test github project.

Test for [using "latest" by default](https://github.com/jaisnan/test-vscode-extension/actions/runs/6101293421/job/16559528415). Related kani-action [here](https://github.com/jaisnan/test-vscode-extension/blob/f0afa13b4542157d27c2aae9cc5a93a5e0dca4a2/.github/workflows/ci.yaml).
Test for [using pinned kani version](https://github.com/jaisnan/test-vscode-extension/actions/runs/6102054828/job/16559681373?pr=2). Related kani-action [here](https://github.com/jaisnan/test-vscode-extension/blob/2e04619cd75bb2692cbdccba791144a000aafd65/.github/workflows/ci.yaml). 

* Is this a refactor change? No

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

celinval commented on .github/workflows/test-action.yml:
```diff
@@ -12,6 +12,11 @@ jobs:
       # you must check out the repository
       - name: Checkout
         uses: actions/checkout@v3
+      - name: Run Kani with older version
```
Nice! One thing you could do is to change the command to actually check if running `kani --version` matches the version you requested.

celinval commented on src/install-kani.sh:
```diff
@@ -0,0 +1,20 @@
+
+# Check exit status for error handling
+if [ $? -ne O ]; then
+    echo "Error installing Kani. Please check if the provided version is correct"
```
Maybe use the github format message:

https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-error-message

jaisnan commented:
Update: I've updated documentation with the features (rendered here) and a test for when an invalid version is passed (like 0.47.0 for example). I've also added checks for the version number to ensure that the latest version is installed by default. But I could not automate parsing the error message that appears when an invalid version is passed in subsequent steps.

celinval reviewed (approved):
Everything looks good, but I would prefer if we can test things a bit better. I'm OK with a follow up PR as long as you manually inspected that the scenarios below work. Thus, I'm approving this one.

Here are the scenarios that I believe should be tested:
  - The default version works and (preferably, check that it is the latest)
     - You could try to run cargo install a second time and make sure the version does not change
     - You could use jq to process the output from [crates.io index](https://crates.io/data-access#crate-index): https://index.crates.io/ka/ni/kani-verifier
  - A specific version works and it matches the requested version
  - An invalid version triggers an error
     - You can use the `outcome` field of the [step context](https://docs.github.com/en/actions/learn-github-actions/contexts#steps-context).
     - Note that `continue-on-error` will continue no matter what, including if the step succeeds.

BTW, I would be OK with testing the installation script separately or test it via github workflow. I prefer the latter though.

celinval reviewed (approved):
Looks good! Thanks @jaisnan! @adpaco-aws do you mind doing a pass on the documentation that Jai added please?

One more thing, should we push this to a branch named v1?

adpaco-aws commented on README.md:
```diff
@@ -2,6 +2,67 @@
+
+NOTE: All the fields provided are optional and have default behaviors when not specified.
+
+`kani-version`
+
+- **Description**: Specifies the Kani version number to use.
+- **Default**: 'latest'
+- **Example-pinned-version**: `'0.33.0'`
+- **Usage**: You can provide a specific version of Kani to use. [Cargo's version specific format is expected for specific versions](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html).
+If omitted, the latest version of `Kani` hosted on [`Kani's crates.io page`](https://crates.io/crates/kani-verifier) will be installed and used.
```
Well done! Although I wonder if we should reduce the space for each parameters by putting all them into a list. For example, taking some inspiration from [this CodeGuru action](https://github.com/marketplace/actions/codeguru-security), we could have this option be:

```
 * `kani-version` (optional, default: `latest`): The Kani version to use. You can provide [...]
```

This would make the documentation feel much lighter.

jaisnan replied:
I've left it as it is for now, I think detail can be good, at least in the beginning. We can always adjust it later on.

adpaco-aws commented on README.md:
```diff
@@ -2,6 +2,67 @@
+
+- **Description**: Specifies the Kani version number to use.
+- **Default**: 'latest'
+- **Example-pinned-version**: `'0.33.0'`
```
What's the purpose of this item? I think it's better to provide examples for the most popular cases in the documentation.

adpaco-aws commented on README.md:
```diff
@@ -2,6 +2,67 @@
+
+Please note that since providing `kani-version` is optional, if the user doesn't provide the version, the action will install the latest version of `Kani` on `crates.io`.
+
+## Example usage in a workflow YAML file:
```
As mentioned earlier, I'd have about 3 examples to illustrate options.

jaisnan replied:
done!

adpaco-aws commented on README.md:
```diff
@@ -2,6 +2,94 @@
+      - name: Run Kani
+        uses: model-checking/kani-github-action@v1.0
+        with:
+          command: 'cargo-kani'
```
Why do you specify `command` on all the examples if it's optional?

adpaco-aws commented on README.md:
```diff
@@ -2,6 +2,94 @@
+- **Default**: false
+- **Usage**: If set to `true`, Kani will enable the experimental PropProof feature for verifying proptest harnesses.
+
+## Kani-version default behavior
+
+Please note that since providing `kani-version` is optional, if the user doesn't provide the version, the action will install the latest version of `Kani` on `crates.io`.
```
I think you're emphasizing this too much.

adpaco-aws commented on README.md:
```diff
@@ -2,6 +2,94 @@
+`kani-version`
+
+- **Description**: Specifies the Kani version number to use.
+- **Default**: 'latest'
```
Can we have backticks around the default values?

adpaco-aws commented on README.md:
```diff
@@ -2,6 +2,94 @@
+
+`kani-version`
+
+- **Description**: Specifies the Kani version number to use.
```
No need to say "Specifies" here nor in the other description. Number is just the way we represent versions, but that's not important. Here's my suggestion:
```suggestion
- **Description**: The Kani version to use.
```

adpaco-aws commented on README.md:
```diff
@@ -2,6 +2,94 @@
+
+- **Description**: Specifies the Kani version number to use.
+- **Default**: 'latest'
+- **Usage**: You can provide a specific version of Kani to use. [Cargo's version specific format is expected for specific versions](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html).
+If omitted, the latest version of `Kani` hosted on [`Kani's crates.io page`](https://crates.io/crates/kani-verifier) will be installed and used.
```
"Usage" seems to rephrase "Description" in most cases. Can we avoid the redundancy? I'd just refer to possible values you can use here (`latest` or a version in SemVer format).

adpaco-aws commented on README.md:
```diff
@@ -2,6 +2,94 @@
+
+- **Description**: Specifies the command to run Kani.
+- **Default**: 'cargo-kani'
+- **Usage**: You can set a custom command to run Kani. For example, this allows you to use a different binary or script for Kani, if needed.
```
I don't understand this. What's `command` really for?
 * Does it only allow `kani` or `cargo-kani`?
 * Does it allow subcommands or is that part of `args`?

adpaco-aws commented on README.md:
```diff
@@ -2,6 +2,94 @@
+
+- **Description**: Experimental feature that allows Kani to verify proptest harnesses using the PropProof feature.
+- **Default**: false
+- **Usage**: If set to `true`, Kani will enable the experimental PropProof feature for verifying proptest harnesses.
```
Can we link to somewhere?

adpaco-aws commented on README.md:
```diff
@@ -2,6 +2,94 @@
+
+Here are a few examples of workflow YAML files for the Kani Github Action
+
+#### Default config which uses the latest version of kani on project
```
Enumerate the examples. And this could just be:
```suggestion
#### Example 1: Default configuration.
```

adpaco-aws commented on README.md:
```diff
@@ -2,6 +2,87 @@
+    steps:
+      - name: Run Kani
+        uses: model-checking/kani-github-action@v1.0
+        with:
+          kani-version: '0.35.0'
+          command: 'cargo-kani'
+          working-directory: './path/to/project'
+          args: '--tests'
+          enable-propproof: true
```
You already demonstrated how to use 3 of these commands before. Can you please only use `args` and `enable-propproof` here?

adpaco-aws commented on README.md:
```diff
@@ -2,6 +2,87 @@
+
+Here are a few examples of workflow YAML files for the Kani Github Action
+
+#### Example 1: Default config which uses the latest version of kani to run `cargo-kani` on project in current directory
```
Wouldn't it be cleaner if we named this `Example 1: Default configuration` and described what it does before the example?

adpaco-aws reviewed (approved):
Documentation looks good to me.

Still think it's a good idea to link mentions of the PropProof feature to our documentation, otherwise they can be confusing.

jaisnan commented:
This change updates the action to add the parameter for `kani-version` which defaults to the value latest.
The default value latest uses the latest version of Kani that is uploaded on `crates.io` unless a specific version is provided by the user, in which case that specific version of Kani will be used in the CI.

Documentation changes [rendered here.](https://github.com/jaisnan/kani-github-action/tree/use-kani-latest-by-default)

The changes were tested in a local setting and a test project here:

    Test for [using "latest" by default](https://github.com/jaisnan/test-vscode-extension/actions/runs/6101293421/job/16559528415). Related kani-github-action job [here](https://github.com/jaisnan/test-vscode-extension/blob/f0afa13b4542157d27c2aae9cc5a93a5e0dca4a2/.github/workflows/ci.yaml).
    Test for [using pinned kani version](https://github.com/jaisnan/test-vscode-extension/actions/runs/6102054828/job/16559681373?pr=2). Related kani-github-action job [here](https://github.com/jaisnan/test-vscode-extension/blob/2e04619cd75bb2692cbdccba791144a000aafd65/.github/workflows/ci.yaml).
